src/mlpack/core/arma_extend/Mat_extra_meat.hpp

```diff
@@ -40,5 +37,5 @@ void Mat<eT>::serialize(Archive& ar, const unsigned int /* version */)
     init_cold();
   }
 
-  ar & make_array(access::rwp(mem), n_elem);
-}
\ No newline at end of file
+  ar & cereal::make_array(access::rwp(mem), n_elem);
+}
```

Comment:
Perhaps this is a good opportunity to rethink the invasive addition of extra functions to the `arma::Mat<eT>` class (ie. beyond the functions that Armadillo already provides).

Adding `arma::Mat<eT>::serialize()` is a brittle approach anyway: if the user includes the Armadillo header before mlpack headers, this type of live-patching won't work. The same applies to `arma::Cube<eT>::serialize()`, `SpMat<eT>::serialize()`, etc.  Furthermore, calling`access::rw()` and `init_cold()` is using internal Armadillo functionality, which is subject to change or removal.

It would be safer and more robust to use an external `serialize()` function, which takes `arma::Mat<eT>` as input.  Both Boost and Cereal provide this option:
* Boost has the "Non Intrusive Version" in https://www.boost.org/doc/libs/1_73_0/libs/serialization/doc/tutorial.html
* Cereal has the "External serialize function": https://uscilab.github.io/cereal/serialization_functions.html

There is no need to write directly into internal Armadillo variables and arrays.  Appropriate constructors can be used.

CC: @rcurtin @zoq

Reply:
Agreed, I see this as an opportunity to fix those things as well, no need to modify Armadillo code to get this done nicely.

Reply:
I wonder why I did it the intrusive way the first time around?  What you've suggested is very clearly better.  Thanks for pointing this out! :+1:

Reply:
@rcurtin Exactly